### ISSUE TITLE
Validate an encoded route key

### ISSUE THREAD

**Issue by ivanvermeyen** [cybercog/laravel-optimus#9]
Hi,

Just FYI: I've put together a validation rule that uses the `resolveRouteBinding()` method on a model to check if a record exists in the database. Sort of like Laravel's `exists` rule but then model based.

https://github.com/codezero-be/laravel-route-key-exists

This works very well with the `OptimusEncodedRouteKey` trait. 👍 

Maybe this can be of use to anyone.

**Comment by a-komarev** [comment]
Hello, @ivanvermeyen.

You package seems useful. Good to have it as separate package as you've done it.

So, it's validating if model exists using encoded key and overwriting this id in request with decoded one?

If I will execute:
```php
$validated = validate([
    'article' => new RouteKeyExists(Article::class)->replace(),
]);
```

This will have actual Article id?
```
dd($validated['article']);
```

So much magic :}

**Comment by ivanvermeyen** [comment]
Indeed, but you'll need extra parentheses:

```php
$validated = validate([
    'article' => (new RouteKeyExists(Article::class))->replace(),
]);
```

I was in doubt if I should add an optional second constructor parameter to enable `replace` or `add`. I think it's slightly less clear, but maybe prettier to look at without those extra parentheses? I think I will add this for flexibility...

```php
$validated = validate([
    'article' => new RouteKeyExists(Article::class, 'article'), // store the original ID in request('article')
]);
```

The real logic for checking if it exists is in the models `resolveRouteBinding()` method, like the one in your trait, so it's not tightly coupled to a single use case. :) All I'm doing is this in essence:

```php
if ( ! $model = $this->model->resolveRouteBinding($value)) {
    return false; // validation fails...
}
```
 And then some checks if I should replace or add the original ID to the request.

**Comment by a-komarev** [comment]
One more possible solution is static factory:
```php
$validated = validate([
    'article' => RouteKeyExists::model(Article::class)->replace(),
]);
```
Or name it as you like: `make` or anything else. Or even this:
```php
$validated = validate([
    'article' => RouteKeyExists::replace(Article::class),
]);
```

**Comment by ivanvermeyen** [comment]
Thanks for the input!

That's also an interesting approach. Looks cleaner than newing up the rule manually.

I've added the static constructor method `model()`. 👍 

```php
RouteKeyExists::model(Article::class)->replace()
```

**Comment by a-komarev** [comment]
Yeah, this one looks good. We could add suggestion for Laravel Optimus users to readme or composer.json that there is optional way to validate if encoded models are exists. But first of all I need to start using them myself. I've never done it in any production projects.
